src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -472,47 +468,192 @@ public synchronized void write(String fileName) throws IOException {
+            case JVM_SIGNATURE_BOOLEAN:
+            case JVM_SIGNATURE_BYTE:
+                size++;
+                break;
+            case JVM_SIGNATURE_CHAR:
+            case JVM_SIGNATURE_SHORT:
+                size += 2;
+                break;
+            case JVM_SIGNATURE_INT:
+            case JVM_SIGNATURE_FLOAT:
+                size += 4;
+                break;
+            case JVM_SIGNATURE_CLASS:
+            case JVM_SIGNATURE_ARRAY:
+                size += OBJ_ID_SIZE;
+                break;
+            case JVM_SIGNATURE_LONG:
+            case JVM_SIGNATURE_DOUBLE:
+                size += 8;
+                break;
+            default:
+                throw new RuntimeException("should not reach here");
+        }
+        return size;
+    }
+
+    private int calculateClassInstanceDumpRecordSize(Instance instance) {
+        Klass reflectedKlass = java_lang_Class.asKlass(instance);
+        // dump instance record only for primitive type Class objects.
+        // all other Class objects are covered by writeClassDumpRecords.
+        if (reflectedKlass == null) {
+            return calculateInstanceDumpRecordSize(instance);
+        }
+        return 0;
+    }
+
+    private int calculateObjectArrayDumpRecordSize(ObjArray array) throws IOException {
+        int headerSize = getArrayHeaderSize(true);
+        final int length = calculateArrayMaxLength(array.getLength(),
+                headerSize,
+                OBJ_ID_SIZE,
+                "Object");
+        return headerSize + length * OBJ_ID_SIZE;
+    }
+
+    private int calculatePrimitiveArrayDumpRecordSize(TypeArray array) throws IOException {
+        int headerSize = getArrayHeaderSize(false);
+        TypeArrayKlass tak = (TypeArrayKlass) array.getKlass();
+        final int type = (int) tak.getElementType();
+        final String typeName = tak.getElementTypeName();
+        final long typeSize = getSizeForType(type);
+        final int length = calculateArrayMaxLength(array.getLength(),
+                                                   headerSize,
+                                                   typeSize,
+                                                   typeName);
+        return headerSize + (int)typeSize * length;
     }
 
     @Override
-    protected void writeHeapRecordEpilogue() throws IOException {
+    protected void writeHeapRecordPrologue(int size) throws IOException {
+        if (size == 0 || (currentSegmentStart > 0)) {
+            return;
+        }
+        // write heap data header
         if (useSegmentedHeapDump) {
-            hprofBufferedOut.exitSegmentMode();
+            out.writeByte((byte)HPROF_HEAP_DUMP_SEGMENT);
+            out.writeInt(0);
+            out.writeInt(size);
+        } else {
+            out.writeByte((byte) (HPROF_HEAP_DUMP));
+            out.writeInt(0);
+            // record position of the length slot
+            currentSegmentStart = fos.getChannel().position();
```

Comment:
The comment before this line is confusing: the `currentSegmentStart` is not the length.
Should this comment be moved after the line 653?

Reply:
Hi Serguei,
Here it gets the file position and record it as the start point of data written, this record would be used later for calculate the data written size which would be written in the file. 
I will update the comment.